docs/design/02-topologies.md

```diff
@@ -0,0 +1,288 @@
+  and when specifying a question for a category, e.g. a `cluster_name` question
+  may be defined in `questions` and then used as `cluster_name` under `all`,
+  however this will not necessarily always need to be the case.
+
+## Notes on `type`
+
+The `type` for a question will serve the dual purpose of validating input and
+then converting this before saving.
+
+For the minimal version of this feature possible type values will probably just
+include `string`, `integer`, and `boolean`. These will be necessary as since we
+are accepting user input we won't know what type we should save a given answer
+as, and we will need non-string types when templating.
+
+Later we can add types with more validation here, e.g. we can have an `ip` type
+which will save the value as a string but validate that it is a valid IP first.
+While these more complicated validations could be specified separately by
+adding an optional regular expression field to question configs, I think
+building these in to Metalware would be better since:
+
+- there will only be a limited number of these more advanced types, and I don't
+  think we need the flexibility of allowing more validations to be specified in
+  the repo;
+
+- having these built in to Metalware means we won't need to duplicate
+  validations between repos or places within repos;
+
+- we may later want to do non-regex validation, e.g. a `dns` type could
+  validate that a particular IP entered gives the expected DNS response (though
+  this particular example sounds like overkill).
+
+
+# `metal configure` commands
+
+The `metal configure` commands each take their specified questions and ask
+them, save the results to a particular file within a new
+`/var/lib/metalware/config/` directory, and perform any other actions
+appropriate to the resource being configured.
+
+## `metal configure all`
+
+Asks the `all` questions; the validated, converted answers will be saved to
+`/var/lib/metalware/config/all.yaml` in the following format:
+
+```yaml
+answers:
+  example_question_identifier: 7
+  another_example: 'foo'
+```
+
+## `metal configure group $GROUP`
+
+Asks for a genders line to be entered first, then similarly to `all` asks the
+`group` questions; the results will then be saved to
+`/var/lib/metalware/config/groups/$GROUP.yaml`
+
+```yaml
+genders: 'somenodes[01-03] somenodes,cluster,all'
+answers:
+  another_example: 'foo'
+```
+
+Note that initially the genders line will just take a line of text, later we
+can add more validation and potentially generate this based on e.g. entering a
+number of nodes in the group and a group name.
+
+Every time `metal configure group` is run we should regenerate
+`/opt/metalware/etc/genders` by (for now at least) simply concatenating all the
+group gender lines together; this way it should always be in sync with the
+configured groups. Aside: it may be worth including a line in the generated
+genders indicating this and that any changes to it directly may not persist.
+
```

Comment:
Sure we can do that, and we definitely will want to if we add specifying genders with more structure later. It will basically amount to the same thing whether we just concatenate the lines together directly or iterate through the groups in a template and include their genders lines, I was just thinking the former would be marginally quicker for the MVP.